src/Binaron.Serializer/Infrastructure/IEnumerableExtension.cs

```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Binaron.Serializer.Infrastructure
+{
+    internal static class IEnumerableExtension
+    {
+        private static ConcurrentDictionary<Type, Action<object, object>> gAdders = new ConcurrentDictionary<Type, Action<object, object>>();
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void Add<T>(this IEnumerable<T> enumerable, T value)
+        {
+            if (!gAdders.TryGetValue(enumerable.GetType(), out Action<object, object> action))

```

Comment:
Maybe use `GetOrAdd` here to stay similar with the other parts of the codebase?

Also, need to think about the unnecessary penalty of doing a concurrent dictionary lookup each time you need to Add an item - we should be able to hoist this out of the Add loop (i.e. try get the action just once and also skip the Add loop if an appropriate Add method isn't found).

Reply:
No problem with the change to GetOrAdd. Will do. 

What about penalty... well, an alternative option is not to use the extension method at all. 
We may replicate the reflection code into every loop that currently uses this extension method. We will lose a little bit because we will search for the method at every deserialization, but won't use a dictionary at every add option. However, that means that we would have 51 copies of this code, if I am not mistaken. :-)

Reply:
_every deserialization means_ to call of `FindMethod` once per object instead of calling once for the type. However, all subsequent add calls would use a local variable. I am not 100% confident how much performance we would gain, as the slowest part here AFAIK is reflection call anyway, it would be interesting to try on one type code and compare.

Reply:
I mean rather than an extension method, let's do something like this:

```C#
    var valueType = Reader.ReadSerializedType(reader);
    var add = TryGetAddAction(...);
    if (add != null)
        while (Reader.ReadEnumerableType(reader) == EnumerableType.HasItem)
        {
            add(...);
        }
```

You still want that dictionary to prevent reflection cost. But the cost of lookup is just once for the whole loop, not once per loop element.

Reply:
Yep, I meant the same. BTW, I was wrong that reflection is the slowest.

The extension method + dictionary is 5-7 times slower than ICollection,
The method like in your code 3-4 times slower.

Reply:
Yep, from one side I like your option better, but even after replacing the invoke with code emitting, the best I get is a 10% improvement. 

Well, it would take quite a few minutes to replace every branch (I tested at one of them), I'll resolve the conversation as I finish.